File: p507e/perfect_number.js
/**
 * @param {number} num
 * @return {boolean}
 */
var checkPerfectNumber = function (num) {
  if (num <= 1) return false;
  let sum = 1;
  for (let i = 2; i * i < num; ++i) {
    if (num % i === 0) sum += i + num / i;
    if (sum > num) return false;
  }
  return sum === num;
};
// TEST
[
  [1, false],
  [6, true],
  [28, true],
  [34, false],
  [496, true],
  [8128, true],
  [33550336, true],
].forEach((t) => {
  const actual = checkPerfectNumber(t[0]);
  console.log('Is', t[0], 'a perfect number? ->', actual);
  console.assert(actual === t[1]);
});

File: p507e/perfect_number.py
class Solution:
    def checkPerfectNumber(self, num: int) -> bool:
        if num <= 1:
            return False
        sum, i = 1, 2
        while i * i < num:
            if num % i == 0:
                sum += i + num / i
            i += 1
            if sum > num:
                return False
        # add only once for perfect square,
        # even without it, it doesn't give false result in the given range
        if i * i == num:
            sum += i
        return sum == num


# TESTS
tests = [
    [1, False],
    [6, True],
    [28, True],
    [34, False],
    [496, True],
    [8128, True],
    [33550336, True],
]
for t in tests:
    sol = Solution()
    actual = sol.checkPerfectNumber(t[0])
    print("Is", t[0], "a perfect number? ->", actual)
    assert actual == t[1]
